cilium/CFP-33716-distributed-packet-tracing.md

```diff
@@ -0,0 +1,586 @@
+#endif
+```
+
+**Pros**
+- Non-competitive storage; since this is not saved directly to a predefined field in the skb struct, there aren’t competing uses for this storage place. It is dedicated directly to the ip_trace_id of the packet being processed.
+
+**Considerations**
+- Must ensure that data is cleared correctly.
+- This is universal storage and not specifically stored in the SKB struct, so it must be cleared when a new packet is processed.
+- At an ingress point, if there is no trace ID stored in a packet, the stored `ip_trace_id` is cleared from the array-map.
+
+### Section 5: Event integration
+
+Include ID as event field for trace and drop structs and the flow message (for Hubble integration):
+
+```c
+struct trace_notify {
+    // Existing struct members
+    ...
+    __u64      ip_trace_id; // new trace ID field
+    // Remaining struct members
+    ...
+};
+```
+```c
+struct drop_notify {
+    // Existing struct members
+    ...
+    __u64      ip_trace_id; // new trace ID field
+    // Remaining struct members
+    ...
+};
+```
+
+```go
+type Flow struct {
+    // Existing struct members
+    ...
+    IpTraceID IpTraceID 
+    `protobuf:"varint,38,opt,name=ip_trace_id,json=ipTraceId,proto3" json:"ip_trace_id,omitempty"` // new trace ID field
+    // Remaining struct members
+    ...
+}
+
+type IpTraceID struct {
+    TraceID      uint64 `protobuf:"varint,1,opt,name=trace_id,json=traceId,proto3" json:"trace_id,omitempty"`
+    IpOptionType uint8  `protobuf:"varint,2,opt,name=ip_option_type,json=ipOptionType,proto3" json:"ip_option_type,omitempty"`
+}
+```
+
+# Benchmark Testing
+
+## Critical User Journeys (CUJs) and Performance Metrics
+
+### Objective:
+These benchmark tests evaluate the performance impact of the new packet tracing feature. The tests focus on measuring packet latency and throughput with and without the tracing feature and custom IP options enabled.
+
+### Key Metrics:
+
+1. **Packet Latency Under High Load**
+   a. **Objective:** Validate that packet latency remains within acceptable limits even under high load conditions.
+   b. **Metric:** Measure the round-trip time (RTT) for packets and ensure it does not exceed a specified threshold (e.g., 1ms average RTT).
+   c. **Scenarios:**
+      i. Baseline (tracing feature disabled, without IP Options)
+      ii. Test (tracing feature disabled, with IP Options)
+      iii. Baseline (tracing feature enabled, without IP Options)
+      iv. Test (tracing feature enabled, with IP Options)
+
+2. **Packet Throughput**
+   a. **Objective:** Assess the maximum throughput the system can handle without degradation in performance.
+   b. **Metric:** Determine the packet sending rate (packets per second, pps) when the rate is set (1000 pps).
+
+3. **Resource Usage**
+   a. **Objective:** Monitor resource usage (CPU, memory) of the control plane with and without the tracing feature under different loads.
+   b. **Metric:** CPU and memory usage.
+
+## Setup
```

Comment:
This experimental section is great -- is there a link to the code?

Reply:
I just pushed it to a repo with the benchmark testing scripts. [Here it is](https://github.com/Bigdelle/packet-tracing-benchmarks/blob/main/benchmark_cpu_usage.sh).